File: 0547-number-of-provinces/0547-number-of-provinces.py
class Solution:

    def __DFSHelp(self, node, neighbors, visited):
        if visited[node] == False:
            visited[node] = True
            for u in neighbors[node]:
                self.__DFSHelp(u, neighbors, visited)

    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        """
        :type isConnected: List[List[int]]
        :rtype: int
        """
        V = range(len(isConnected))
        neighbors={}
        visited={}

        for v in V:
            neighbors[v]=[]
            visited[v]=False

        for v in V:
            for i, u in enumerate(isConnected[v]):
                if u == 1:
                    neighbors[v].append(i)
        
        count = 0
        for v in V:
            if not visited[v]:
                count+=1
                self.__DFSHelp(v, neighbors, visited)

        return count
        

File: graphs-dfs/0547-number-of-provinces/0547-number-of-provinces.c
#include <stdio.h>
#include <stdbool.h>

void dfsHelp(int node, bool *aVisited, int** aIsConnected, int aSize){
    if(aVisited[node]==true){return;}
    else{
        aVisited[node]=true;
        for(int i=0; i< aSize; i++){
            if(aIsConnected[node][i]==1){
                dfsHelp(i, aVisited, aIsConnected, aSize);
            }
        }
        return;
    }
}

int findCircleNum(int** isConnected, int isConnectedSize, int* isConnectedColSize){

    bool visited[isConnectedSize];
    for(int i=0; i<isConnectedSize; i++){
        visited[i]=false;
    }
    int nCircle=0;
    for(int i=0; i<isConnectedSize; i++){
        if(visited[i]==false){
            nCircle++;
            dfsHelp(i, visited, isConnected, isConnectedSize);
        }
    }
    return nCircle;
}



int main(){
    // int n=3;
    // int *myIsConnected[3] = { // parameter of type int**: https://stackoverflow.com/questions/41647445/incompatible-pointer-types-passing-int55-to-parameter-of-type-int
    //     (int[]){1,1,0}, 
    //     (int[]){1,1,0}, 
    //     (int[]){0,0,1}
    // };
    int n=3;
    int *myIsConnected[3] = {
        (int[]) {1,0,0},
        (int[]) {0,1,0},
        (int[]) {0,0,1}
    };
    int numProvinces = findCircleNum(myIsConnected, n, myIsConnected[0]);
    printf("num provinces : %d\n", numProvinces);
}